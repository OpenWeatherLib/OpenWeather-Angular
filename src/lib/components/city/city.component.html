<mat-card class="ga-card" *ngIf="city">
  <mat-card-header>
    <img mat-card-avatar class="ga-card-header-image" src="{{cityPictureUrl}}" alt="City pic">
    <mat-card-title>City</mat-card-title>
    <mat-card-subtitle>{{city?.name}}</mat-card-subtitle>
  </mat-card-header>
  <img mat-card-image src="{{cityPictureUrl}}" alt="City pic">
  <mat-card-content>
    <p>Country: {{city?.country}}</p>
    <p>Population: {{city?.population}}</p>
    <p>Coordinates: Lat: {{city?.coord?.lat}}, Lon: {{city?.coord?.lon}}</p>
  </mat-card-content>
</mat-card>
