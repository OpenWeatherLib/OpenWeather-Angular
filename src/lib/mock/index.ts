export * from "./methods.mock";
export * from "./values.mock";
