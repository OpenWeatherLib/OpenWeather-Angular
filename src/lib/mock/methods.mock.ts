declare var jasmine;

export const substitute = <T>(type: new (...x: any[]) => T): jasmine.SpyObj<T> =>
    jasmine.createSpyObj(
        type.prototype.constructor.name,
        Object.getOwnPropertyNames(type.prototype).filter(key => typeof type.prototype[key] === "function"));
