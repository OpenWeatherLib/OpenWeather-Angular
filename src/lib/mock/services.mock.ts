declare var jasmine;

export default class MockServices {
    static substitute<T>(type: { new(...x): T }): jasmine.SpyObj<T> {
        const obj = type.prototype;

        const objPropertyNames = Object.getOwnPropertyNames(obj);
        const methods = objPropertyNames.filter(key => typeof obj[key] === "function");

        return jasmine.createSpyObj(
            obj.constructor.name,
            methods
        );
    }
}
