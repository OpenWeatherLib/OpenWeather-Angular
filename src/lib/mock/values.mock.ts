import {
    AddressComponent,
    City,
    City2,
    Coordinates2,
    Geometry,
    UvIndex,
    Viewport,
    WeatherCurrent,
    WeatherForecast,
    WeatherForecastPart,
    WeatherPart
} from "@lib/models";

export default class MockValues {
    static addressComponentCity(): AddressComponent {
        return {
            short_name: "Nuremberg",
            types: [
                "locality",
                "political"
            ],
            long_name: "Nuremberg, DE"
        };
    }

    static addressComponentCountry(): AddressComponent {
        return {
            short_name: "DE",
            types: [
                "country",
                "political"
            ],
            long_name: "Germany"
        };
    }

    static apiKey(): string {
        return "fdsje423rfsdu";
    }

    static city(): City {
        const city = new City();
        city.id = 420;
        city.name = "Nuremberg";
        city.country = "DE";
        city.population = 499523;
        city.coord = { lat: 45.3452, lon: 23.543 };
        return city;
    }

    static city2(): City2 {
        return {
            address_components: [
                this.addressComponentCity(),
                this.addressComponentCountry()
            ],
            geometry: this.geometry(),
            types: [
                "locality",
                "political"
            ]
        };
    }

    static coordinatesNE(): Coordinates2 {
        return {
            lat: 49.5730438232,
            lng: 11.3236637115
        };
    }

    static coordinatesSW(): Coordinates2 {
        return {
            lat: 49.3166618347,
            lng: 10.9793968201
        };
    }

    static geometry(): Geometry {
        return {
            location_type: "APPROXIMATE",
            viewport: this.viewport(),
            location: this.location()
        };
    }

    static location(): Coordinates2 {
        return {
            lat: 49.45421,
            lng: 11.07752
        };
    }

    static uvIndex(): UvIndex {
        return {
            lat: 45.3452,
            lon: 23.543,
            date_iso: "2018-06-29 03:00:00",
            date: 1530219600,
            value: 4.20
        };
    }

    static viewport(): Viewport {
        return {
            northeast: this.coordinatesNE(),
            southwest: this.coordinatesSW()
        };
    }

    static weatherCurrent(): WeatherCurrent {
        return {} as WeatherCurrent;
    }

    static weatherForecast(): WeatherForecast {
        return {} as WeatherForecast;
    }

    static weatherForecastPart(): WeatherForecastPart {
        return {} as WeatherForecastPart;
    }

    static weatherPart(): WeatherPart {
        return {} as WeatherPart;
    }
}
