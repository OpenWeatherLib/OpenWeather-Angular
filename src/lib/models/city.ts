import { GeoLocation } from "./geo-location";
import { classJsonKey } from "../decorator";

@classJsonKey("", "")
export class City {
  id: number = 0;
  name: string = "";
  country: string = "";
  population: number = 0;
  geoLocation: GeoLocation = new GeoLocation();

  isDefault(): boolean {
    return this.id === 0 && this.name === "" && this.country === "" && this.population === 0 && this.geoLocation.isDefault();
  }
}
