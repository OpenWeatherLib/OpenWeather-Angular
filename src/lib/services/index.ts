export * from "./api.service";
export * from "./image.service";
export * from "./open-weather.service";
