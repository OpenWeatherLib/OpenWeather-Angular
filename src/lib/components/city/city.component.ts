import { Component, Input, OnInit } from "@angular/core";

import { BaseComponent } from "@lib/components/base-component/base.component";
import { City } from "@lib/models";
import { ImageService, OpenWeatherService } from "@lib/services";

@Component({
  selector: "ga-city",
  templateUrl: "./city.component.html",
  styleUrls: ["./city.component.scss"]
})
export class CityComponent extends BaseComponent implements OnInit {

  @Input() initialCityName: string = String().empty;

  city: City = null;
  cityPictureUrl: string = String().empty;

  constructor(private readonly imageService: ImageService,
    private readonly openWeatherService: OpenWeatherService) {
    super();
  }

  ngOnInit() {
    this.openWeatherService.loadCityData(this.initialCityName);

    this.registerSubscription(
      this.openWeatherService.city()
        .subscribe(city => {
          if (city) {
            this.city = city;
            this.imageService.receiveImagePictureUrl(this.city.name);
          }
        }));

    this.registerSubscription(
      this.imageService.cityPictureUrl()
        .subscribe(url => {
          this.cityPictureUrl = url;
        }));
  }
}
