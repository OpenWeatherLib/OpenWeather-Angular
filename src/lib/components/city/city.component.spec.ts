import { TestBed, async } from "@angular/core/testing";
import { of } from "rxjs";

import { substitute } from "@lib/mock";

import { ImageService, OpenWeatherService } from "@lib/services";

import { CityComponent } from "./city.component";

describe("CityComponent", () => {
  let classToTest: CityComponent;

  const imageServiceMock = substitute(ImageService);
  const openWeatherServiceMock = substitute(OpenWeatherService);

  const serviceMockList: any[] = [
    imageServiceMock,
    openWeatherServiceMock
  ];

  beforeEach(() => {
    imageServiceMock.cityPictureUrl.and.returnValue(of(null));
    openWeatherServiceMock.city.and.returnValue(of(null));

    TestBed.configureTestingModule({
      declarations: [
        CityComponent
      ],
      providers: [
        { provide: ImageService, useValue: imageServiceMock },
        { provide: OpenWeatherService, useValue: openWeatherServiceMock }
      ]
    }).overrideTemplate(CityComponent, "<div></div>").compileComponents();

    const fixture = TestBed.createComponent(CityComponent);
    classToTest = fixture.debugElement.componentInstance;
  });

  afterEach(() => {
    serviceMockList.forEach(serviceMock => {
      for (const propertyName in serviceMock) {
        if (serviceMock.hasOwnProperty(propertyName)) {
          serviceMock[propertyName].calls.reset();
          serviceMock[propertyName].and.stub();
        }
      }
    });
  });

  it("should create the app", async(() => {
    // Assert
    expect(classToTest).toBeTruthy();
  }));
});
