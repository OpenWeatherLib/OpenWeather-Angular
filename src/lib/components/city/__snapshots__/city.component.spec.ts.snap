// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityComponent ngOnInit city$ 1`] = `
Object {
  "coord": Object {
    "lat": 49.460983,
    "lon": 11.061859,
  },
  "country": "DE",
  "id": 0,
  "name": "Nuremberg",
  "population": 518365,
}
`;
